scripts/pre_commit_linter.py

```diff
@@ -1049,6 +1049,86 @@ def _check_docstrings(all_files):
     return summary_messages
 
 
+def _check_controller_class_names(all_files):
+    """This function checks that all controller class names end with
+    either 'Handler' or 'Page'.
+    """
+    print 'Starting controller class names check...'
+    print '----------------------------------------'
+    summary_messages = []
+    controllers_path = 'core/controllers*'
+    files_to_check = [
+        filename for filename in all_files if not
+        any(fnmatch.fnmatch(filename, pattern) for pattern in EXCLUDED_PATHS)
+        and filename.endswith('.py') and not filename.endswith('_test.py')
+        and fnmatch.fnmatch(filename, controllers_path)]
+    failed = False
+    for filename in files_to_check:
+        with open(filename, 'r') as f:
+            file_content = f.readlines()
+            file_length = len(file_content)
+            class_name = None
+            previous_class_name = None
+            super_class_name = None
+            previous_super_class_name = None
+            line_num_of_class = None
+            class_name_passed = True
+            function_name = None
+            for line_num in range(file_length):
```

Comment:
This string-based approach feels a bit fragile. E.g. there are baked-in assumptions (I think?) that self.render_template isn't aliased, all "def" stuff occurs in classes, etc. In addition the code is a bit long and hard to follow, and it's not straightforward to verify that it's fully correct and comprehensive.

Instead, how about an AST/object-based approach like the following?

- For each controller file, find all classes in it that are subclasses of BaseHandler. We already do something similar in our interaction registry, you can use similar methods.
- For each of those classes, introspect what methods it has. Also I think there is a class-level constant that determines whether the return type is HTML or JSON, you can use that to distinguish between the two types of classes.
- Use `class.__name__` to get the name info and make sure it's correct.

I think this will lead to shorter and more understandable code, and it will be more maintainable in the long run than parsing through every line. Does that sound good?

Reply:
@lilithxxx ping?

Reply:
Hi @apb7 @seanlip I am having a problem. The following is the traceback:
``` File "scripts/pre_commit_linter.py", line 1656, in main
    controller_class_name_messages = _check_controller_class_names(all_files)
  File "scripts/pre_commit_linter.py", line 1072, in _check_controller_class_names
    module = loader.find_module(name).load_module(name)
  File "/usr/lib/python2.7/pkgutil.py", line 243, in load_module
    mod = imp.load_module(fullname, self.file, self.filename, self.etc)
  File "/home/jude/foss/oppia/core/controllers/base.py", line 51, in <module>
    'Text used to encrypt CSRF tokens.', DEFAULT_CSRF_SECRET)
  File "/home/jude/foss/oppia/core/domain/config_domain.py", line 114, in __init__
    raise Exception('Property with name %s already exists' % name)
Exception: Property with name oppia_csrf_secret already exists
```
This happens when iterating through all modules in core/controllers

Reply:
This seems like something you would need to debug. Try and read through the relevant lines of code in order to figure out what is causing the error to occur. If you can diagnose the root cause of the issue then you are halfway towards fixing it.

Also re this other comment:

> I found a constant called 'GET_HANDLER_ERROR_RETURN_TYPE' which I think tells what the class returns. But I found many classes like in [here](https://github.com/oppia/oppia/blob/d2cfa460cabd22dafc3b06cdc16da0df43cd8482/core/controllers/base.py#L484) where according to me it should end with Handler but the constant is set to html. There are many more examples like this. Maybe I picked the wrong constant but if I didn't there are many things that needs to be changed.

I think the particular class you linked to is kind of a special case (it's the controller that handles 404 errors), though I agree that it should probably end with *Page instead. If you could send links to the others, then I could try and give you advice on what to do with them!

Reply:
There are a lot of classes and I am thinking the constant needs to be changed accordingly. 
These are the classes:
```
core/controllers/admin.py --> Line 324: Please ensure that the name of this class ends with 'Page'
core/controllers/base.py --> Line 484: Please ensure that the name of this class ends with 'Page'
core/controllers/classifier.py --> Line 90: Please ensure that the name of this class ends with 'Page'
core/controllers/classifier.py --> Line 138: Please ensure that the name of this class ends with 'Page'
core/controllers/collection_editor.py --> Line 47: Please ensure that the name of this class ends with 'Page'
core/controllers/collection_editor.py --> Line 168: Please ensure that the name of this class ends with 'Page'
core/controllers/collection_editor.py --> Line 205: Please ensure that the name of this class ends with 'Page'
core/controllers/collection_editor.py --> Line 234: Please ensure that the name of this class ends with 'Page'
core/controllers/concept_card_viewer.py --> Line 23: Please ensure that the name of this class ends with 'Page'
core/controllers/creator_dashboard.py --> Line 373: Please ensure that the name of this class ends with 'Page'
core/controllers/creator_dashboard.py --> Line 391: Please ensure that the name of this class ends with 'Page'
core/controllers/creator_dashboard.py --> Line 407: Please ensure that the name of this class ends with 'Page'
core/controllers/cron.py --> Line 42: Please ensure that the name of this class ends with 'Page'
core/controllers/cron.py --> Line 81: Please ensure that the name of this class ends with 'Page'
core/controllers/cron.py --> Line 91: Please ensure that the name of this class ends with 'Page'
core/controllers/cron.py --> Line 102: Please ensure that the name of this class ends with 'Page'
core/controllers/cron.py --> Line 112: Please ensure that the name of this class ends with 'Page'
core/controllers/cron.py --> Line 193: Please ensure that the name of this class ends with 'Page'
core/controllers/cron.py --> Line 209: Please ensure that the name of this class ends with 'Page'
core/controllers/editor.py --> Line 86: Please ensure that the name of this class ends with 'Page'
core/controllers/editor.py --> Line 112: Please ensure that the name of this class ends with 'Page'
core/controllers/email_dashboard.py --> Line 40: Please ensure that the name of this class ends with 'Page'
core/controllers/email_dashboard.py --> Line 126: Please ensure that the name of this class ends with 'Page'
core/controllers/email_dashboard.py --> Line 190: Please ensure that the name of this class ends with 'Page'
core/controllers/email_dashboard.py --> Line 208: Please ensure that the name of this class ends with 'Page'
core/controllers/feedback.py --> Line 64: Please ensure that the name of this class ends with 'Page'
core/controllers/feedback.py --> Line 169: Please ensure that the name of this class ends with 'Page'
core/controllers/feedback.py --> Line 184: Please ensure that the name of this class ends with 'Page'
core/controllers/feedback.py --> Line 251: Please ensure that the name of this class ends with 'Page'
core/controllers/incoming_emails.py --> Line 28: Please ensure that the name of this class ends with 'Page'
core/controllers/learner_dashboard.py --> Line 123: Please ensure that the name of this class ends with 'Page'
core/controllers/learner_dashboard.py --> Line 145: Please ensure that the name of this class ends with 'Page'
core/controllers/learner_playlist.py --> Line 24: Please ensure that the name of this class ends with 'Page'
core/controllers/library.py --> Line 178: Please ensure that the name of this class ends with 'Page'
core/controllers/profile.py --> Line 337: Please ensure that the name of this class ends with 'Page'
core/controllers/profile.py --> Line 357: Please ensure that the name of this class ends with 'Page'
core/controllers/question_editor.py --> Line 28: Please ensure that the name of this class ends with 'Page'
core/controllers/question_editor.py --> Line 65: Please ensure that the name of this class ends with 'Page'
core/controllers/question_editor.py --> Line 95: Please ensure that the name of this class ends with 'Page'
core/controllers/reader.py --> Line 313: Please ensure that the name of this class ends with 'Page'
core/controllers/reader.py --> Line 345: Please ensure that the name of this class ends with 'Page'
core/controllers/reader.py --> Line 530: Please ensure that the name of this class ends with 'Page'
core/controllers/reader.py --> Line 583: Please ensure that the name of this class ends with 'Page'
core/controllers/reader.py --> Line 634: Please ensure that the name of this class ends with 'Page'
core/controllers/reader.py --> Line 667: Please ensure that the name of this class ends with 'Page'
core/controllers/reader.py --> Line 687: Please ensure that the name of this class ends with 'Page'
core/controllers/reader.py --> Line 703: Please ensure that the name of this class ends with 'Page'
core/controllers/reader.py --> Line 730: Please ensure that the name of this class ends with 'Page'
core/controllers/reader.py --> Line 754: Please ensure that the name of this class ends with 'Page'
core/controllers/reader.py --> Line 773: Please ensure that the name of this class ends with 'Page'
core/controllers/reader.py --> Line 791: Please ensure that the name of this class ends with 'Page'
core/controllers/reader.py --> Line 845: Please ensure that the name of this class ends with 'Page'
core/controllers/reader.py --> Line 887: Please ensure that the name of this class ends with 'Page'
core/controllers/reader.py --> Line 1006: Please ensure that the name of this class ends with 'Page'
core/controllers/resources.py --> Line 26: Please ensure that the name of this class ends with 'Page'
core/controllers/resources.py --> Line 42: Please ensure that the name of this class ends with 'Page'
core/controllers/resources.py --> Line 75: Please ensure that the name of this class ends with 'Page'
core/controllers/skill_editor.py --> Line 94: Please ensure that the name of this class ends with 'Page'
core/controllers/skill_editor.py --> Line 121: Please ensure that the name of this class ends with 'Page'
core/controllers/skill_editor.py --> Line 189: Please ensure that the name of this class ends with 'Page'
core/controllers/story_editor.py --> Line 58: Please ensure that the name of this class ends with 'Page'
core/controllers/subscriptions.py --> Line 23: Please ensure that the name of this class ends with 'Page'
core/controllers/subscriptions.py --> Line 34: Please ensure that the name of this class ends with 'Page'
core/controllers/suggestion.py --> Line 28: Please ensure that the name of this class ends with 'Page'
core/controllers/suggestion.py --> Line 43: Please ensure that the name of this class ends with 'Page'
core/controllers/suggestion.py --> Line 86: Please ensure that the name of this class ends with 'Page'
core/controllers/suggestion.py --> Line 132: Please ensure that the name of this class ends with 'Page'
core/controllers/tasks.py --> Line 32: Please ensure that the name of this class ends with 'Page'
core/controllers/tasks.py --> Line 72: Please ensure that the name of this class ends with 'Page'
core/controllers/tasks.py --> Line 90: Please ensure that the name of this class ends with 'Page'
core/controllers/tasks.py --> Line 118: Please ensure that the name of this class ends with 'Page'
core/controllers/tasks.py --> Line 143: Please ensure that the name of this class ends with 'Page'
core/controllers/topic_editor.py --> Line 40: Please ensure that the name of this class ends with 'Page'
core/controllers/topic_editor.py --> Line 95: Please ensure that the name of this class ends with 'Page'
core/controllers/topic_editor.py --> Line 176: Please ensure that the name of this class ends with 'Page'
core/controllers/topic_editor.py --> Line 217: Please ensure that the name of this class ends with 'Page'
core/controllers/topic_editor.py --> Line 326: Please ensure that the name of this class ends with 'Page'
core/controllers/topic_editor.py --> Line 355: Please ensure that the name of this class ends with 'Page'
core/controllers/topic_editor.py --> Line 380: Please ensure that the name of this class ends with 'Page'
core/controllers/topic_viewer.py --> Line 45: Please ensure that the name of this class ends with 'Page'
core/controllers/topics_and_skills_dashboard.py --> Line 118: Please ensure that the name of this class ends with 'Page'
core/controllers/topics_and_skills_dashboard.py --> Line 138: Please ensure that the name of this class ends with 'Page'
core/controllers/translator.py --> Line 52: Please ensure that the name of this class ends with 'Page'
core/controllers/translator.py --> Line 142: Please ensure that the name of this class ends with 'Page'
```

Reply:
And regarding the above error it occurs [here](https://github.com/oppia/oppia/blob/develop/core/domain/config_domain.py#L114). Is there any way to clear all ```ConfigProperty``` before running the script or commenting out the line(maybe?)

Reply:
@seanlip And regarding the above error it occurs here. Is there any way to clear all ConfigProperty before running the script or commenting out the line(maybe?)

Reply:
Hi @lilithxxx, sorry for the delay on replying to this. 

Regarding the error, I think it arises because you are manually loading each module repeatedly, but the ConfigProperty can only be initialized once. So, here is another approach. Perhaps you can iterate through the list of URL tuples in main.py instead and get their handler classes from there? main.py already imports those handlers so the normal import flow should work fine.

Regarding the list of files, thank you for generating it. I agree that some further cleanup is needed as there are a bunch of other cases. Comments below for the first part -- I think it should be possible to generalize these rules to the rest. If you come across cases that need additional direction, please let me know and I can suggest what to do.

Thanks!

```
core/controllers/admin.py --> Line 324: This generates a file for download, which seems like it should be treated as a third "return type". So, we need to design some kind of error return type for handlers that generate files to be downloaded, that's analogous to json -- and maybe introduce a render_downloadable_file() method in the base handler to deal with stuff like this.
core/controllers/base.py --> Line 484: I am confused about this one. It's not referring to a handler?
core/controllers/classifier.py --> Line 90: Our checks should ensure that any controller without a get() method should end with *Handler.
core/controllers/classifier.py --> Line 138: Ditto (no get() defined).
core/controllers/collection_editor.py --> Line 47: Ditto (no get() defined).
core/controllers/collection_editor.py --> Line 168: Ditto (no get() defined).
core/controllers/collection_editor.py --> Line 205: Ditto (no get() defined).
core/controllers/collection_editor.py --> Line 234: This should be a JSON handler and we should add the relevant ERROR... constant to it.
core/controllers/concept_card_viewer.py --> Line 23: Ditto (add relevant constant).
core/controllers/creator_dashboard.py --> Line 373: Ditto (no get() defined).
core/controllers/creator_dashboard.py --> Line 391: Ditto (no get() defined).
core/controllers/creator_dashboard.py --> Line 407: Ditto (no get() defined).
core/controllers/cron.py --> Line 42 etc.: I think you can have a rule that says "if the handler is in cron.py then it should end with Handler".
...
```

Reply:
Hi @seanlip sorry for such a delay in update! I have made the changes as you said and just a few questions:
> core/controllers/admin.py --> Line 324: This generates a file for download, which seems like it should be treated as a third "return type". So, we need to design some kind of error return type for handlers that generate files to be downloaded, that's analogous to json -- and maybe introduce a render_downloadable_file() method in the base handler to deal with stuff like this.

I have made the function but in the end what will the class name end with?

There are also two classes in `main.py` which are extended from `base` [here](https://github.com/oppia/oppia/blob/develop/main.py#L79) and [here](https://github.com/oppia/oppia/blob/develop/main.py#L88). Should these be taken into account for the test and if so what should their name end with?

One final question is that you suggested to iterate through the URL tuples in `main` but there seems to be a linting error:
```
Redefining name 'main' from outer scope (line 1608) (redefined-outer-name)
```
I think this is due to the fact that there exists a function called `main` inside `pre_commit_linter` and hence there seems some kind of problem when I try importing `main.py`

Reply:
(1) Perhaps FileDownloader?

(2) Yep, take them into account for the test -- it's generally better if there's no exceptions. HomePageRedirectHandler is "html", I think; redirection doesn't make sense for JSON responses. WarmupHandler could be either.

(3) You could use the "from X import Y as Z" syntax. But maybe we should actually have that as a backend test in core/controllers/base_test.py rather than something in pre commit linter, because importing main implies importing the whole application and I don't think that's something you want to do in the linter script. @apb7 what do you think about this suggestion?

Reply:
@seanlip made the changes!